fix: compress option and accept/content encoding header edge cases

tsibley opened the pull request:
I ran into some edge cases with the caching behaviour of the `compress` option and with the handling of `Accept-Encoding` and `Content-Encoding`. I put details in the commit messages and tried to address each issue separately (though they're of a family) for the most clarity, with tests and then fixes. The quick summary is:

- Make `Accept-Encoding` handling internally consistent
- Prevent invalid `Content-Encoding: null` on some responses
- Vary caching on the `compress` option to avoid mixing up responses between requests with `compress` enabled vs. disabled
- Cleanup a few small things I noticed in passing

Tests pass locally for me when I run `npm run test`.

I'm not deeply involved in the Node ecosystem, so let me know if I've missed anything "obvious" about PR/testing/etc here.

wraithgar commented:
Those deleted tests sure look scary at first glance don't they?  Good catch finding those dupes though.

Don't worry about the failing node10 CI that's not due to this PR it's an existing issue.

At a glance this looks more correct but I think @nlf should weigh in as he was the last one to do a real deep dive on caching and headers.

nlf commented:
ok, i just want to make sure i fully understand the changes you've made here so i'm going to restate them and you can let me know if i got it wrong:

1. always store the request's `accept-encoding` header
2. store the value of the `compress` option as part of the cached metadata
3. only store `content-encoding` when `{ compress: false }` is in the options, and only when the response header actually has a value
4. do not serve cached content if the `compress` option has a different value than the cached metadata

is that correct?

nlf commented:
also, what's the specific bug you were seeing manifest? receiving uncompressed results for cached responses with `accept-encoding: gzip`?

tsibley commented:
> 1. always store the request's `accept-encoding` header
> 
> 2. store the value of the `compress` option as part of the cached metadata

Correct.

> 3. only store `content-encoding` when `{ compress: false }` is in the options, and only when the response header actually has a value

Correct.

I think even better behaviour would be to extend the above to also store `content-encoding` when `{ compress: true }` but minipass-fetch wasn't able to decode the response body, but I didn't see a way to reliably detect that currently without patching minipass-fetch too. So the current behaviour here assumes that minipass-fetch is always successful at decoding.

> 4. do not serve cached content if the `compress` option has a different value than the cached metadata

Correct.

> also, what's the specific bug you were seeing manifest? receiving uncompressed results for cached responses with `accept-encoding: gzip`?

The initial bug I ran into in my code was with requests that had an `accept-encoding: gzip` and `compress: false`. I was seeing `content-encoding: null` in responses returned from the cache after revalidation (triggered by `cache: 'no-cache'`). From an empty cache, the sequence would be:

1. Miss, populates cache, response ok.
2. Revalidates, updates cache, response ok.
3. Revalidates, updates cache, response with `content-encoding: null`.

I can write a specific separate test for this. I thought I had during my debugging at one point but I see now it didn't get included in the branch.

While investigating that, I noticed the issue of mixing up stored responses between requests with `compress` enabled vs. disabled. I expect this would also crop up in my codebase, but I hadn't noticed it there yet.

tsibley commented:
> > 3. only store `content-encoding` when `{ compress: false }` is in the options, and only when the response header actually has a value

Hmm. Looking at this again, I realized that I introduced the opposite bug. orz

Consider this test suite: https://gist.github.com/tsibley/f23d0c77291bcbe56b951092903f2e89

I'll look to fix this.

tsibley commented:
Ok, I've reworked the changes here after further thought about how they interact and what they're trying to solve. I've added a whole separate test file which more clearly (to me, at least) demonstrates the broken state and the fixed state. Feeling better about this rework being a simpler fix, as it removes special casing of headers re: the compress option.

tsibley commented on lib/cache/entry.js:
```diff
@@ -194,6 +189,7 @@ class CacheEntry {
       this[_request] = new Request(this.entry.metadata.url, {
         method: 'GET',
         headers: this.entry.metadata.reqHeaders,
+        ...(this.entry.metadata.options || {}),
```
Worth noting: the `|| {}` is necessary to handle existing caches with the new code since existing cache entries won't have a `metadata.options` stored. The result is that existing caches will assume previously cached requests were made with `{compress: true}` (because it's the default of `Request`) and match new requests also made with `{compress: true}`. If the previously cached requests were actually made with `{compress: false}`, then a stored compressed response may be returned to the new `{compress: true}` request. This is incorrect behaviour, but also the existing behaviour before the new code in this PR, so :man_shrugging:. The fix in this edge case is to delete the existing cache and let it repopulate.

AFAICT, there's no way to determine what the actual `compress` option was for these existing entries, so some value must be assumed. We could maybe instead propagate an "unknown" value forward into the `satisfies()` check, but all this would let us do is treat all existing cache entries as misses for new requests. It's not clear to me that's any better.

nlf replied:
since `this.entry.metadata` _does_ exist, it's safe to leave off the `|| {}` because `undefined` is allowed in an object spread, it just doesn't add anything.

```
> var entry = { metadata: {} }
> { ...entry.metadata.options }
{}
```

tsibley replied:
Ah, didn't know that about the spread operator! I thought it would throw a TypeError or something. Will fix.

tsibley commented:
@nlf @wraithgar My rework seems to maybe have stalled things. Did time to review just evaporate, or are there as-of-yet-unspoken concerns with the changes?

tsibley commented:
@nlf Ok, thank you! :pray: No worries about getting busy. Just wanted to check in and make sure there wasn't something I could do to unblock it.

nlf commented on lib/cache/policy.js:
```diff
@@ -97,6 +98,9 @@ class CachePolicy {
     if (this.request.headers.host !== _req.headers.host)
       return false
 
+    if (this.request.compress !== request.compress)
```
since you've added the `compress` option as part of the result in the `requestObject()` function above, this should probably compare to `_req.compress` instead of `request.compress` for consistency

tsibley replied:
Yes, indeed. Will fix.

nlf commented:
ok, so i think you've taken the safest route possible here varying responses based on the `compress` option. it seems very reasonable to me. it does mean we'll potentially not serve things from cache that we could, but it also means anything we do handle is of the correct encoding which is far more important than the occasional cache miss that could've been a hit.

i do think this deserves a closer look to try to determine how we're ending up with `content-encoding: null` in entries in the first place. i think it likely is the code that blindly copies the header when `compress` is `false`, which should check to make sure the header is present before copying. however, we should _not_ store the `content-encoding` if we've stored a decompressed entry as it is a lie and we know that. it should only be stored if `compress` is `false` and the header exists on the response. arguably, minipass-fetch is what should be removing that header from the response when it knows it'll be decompressing the body, but we can certainly guard against it here too. sending a decompressed response with a `content-encoding` header signaling otherwise is flat out wrong.

also, since we know that the failure state here is we end up with a cache entry that has `content-encoding: null` in its headers, we should extend the `validateEntry` function [here](https://github.com/npm/make-fetch-happen/blob/main/lib/cache/entry.js#L145-L151) such that it considers any entry in that state to be invalid, which will remove them from the index and clean up the bad entries on the fly.

and lastly, while i appreciate that you've written tests for these things, i think the DRY approach there leaves the tests in a state that each one doesn't tell a full story. for tests, it's preferable to keep as much context as is reasonable within the test. persistent nocks don't tell us how many times we expect the actual network request to be made. a wrapper function for all of the assertions takes a lot of context away from the test and requires us to scroll down and parse the wrapper function against the inputs. i'd love it if we can unroll those abstractions so that each test tells a full end-to-end story about exactly what takes place and when.

tsibley commented:
> i do think this deserves a closer look to try to determine how we're ending up with `content-encoding: null` in entries in the first place. i think it likely is the code that blindly copies the header when `compress` is `false`, which should check to make sure the header is present before copying.

Indeed. I addressed that in my previous version of this PR (see https://github.com/npm/make-fetch-happen/commit/ef415a9e4124c82165dcdd2413a9b0071ddfe385), but later I concluded the Content-Encoding handling was really a red herring for the underlying issue of not varying on `compress`.

> however, we should _not_ store the `content-encoding` if we've stored a decompressed entry as it is a lie and we know that. it should only be stored if `compress` is `false` and the header exists on the response.

I agree on principle, but not sure it's that simple in practice, :point_down: 

> arguably, minipass-fetch is what should be removing that header from the response when it knows it'll be decompressing the body, but we can certainly guard against it here too. sending a decompressed response with a `content-encoding` header signaling otherwise is flat out wrong.

I agree that minipass-fetch is the ideal place to remove the Content-Encoding header and that barring that make-fetch-happen could almost always detect when it's a lie. However, minipass-fetch's behaviour of decompressing the body but leaving the Content-Encoding header matches that of node-fetch. That behaviour compat may be important for minipass-fetch and/or make-fetch-happen as they aim to be drop-in replacements for node-fetch?

Notably, the browser `fetch()` (at least in recent Firefox) does the right thing and removes the Content-Encoding header if it was able to decode the body (and leaves it if not).

> also, since we know that the failure state here is we end up with a cache entry that has `content-encoding: null` in its headers, we should extend the `validateEntry` function [here](https://github.com/npm/make-fetch-happen/blob/main/lib/cache/entry.js#L145-L151) such that it considers any entry in that state to be invalid, which will remove them from the index and clean up the bad entries on the fly.

Ah, great point! Will add this.

I guess the same mechanism could be used to prevent the [incorrect behaviour continuing to happen for existing caches](https://github.com/npm/make-fetch-happen/pull/65#discussion_r762171534), but the effect would be to wipe the cache and maybe that's too dramatic?

> and lastly, while i appreciate that you've written tests for these things, i think the DRY approach there leaves the tests in a state that each one doesn't tell a full story. for tests, it's preferable to keep as much context as is reasonable within the test. persistent nocks don't tell us how many times we expect the actual network request to be made. a wrapper function for all of the assertions takes a lot of context away from the test and requires us to scroll down and parse the wrapper function against the inputs. i'd love it if we can unroll those abstractions so that each test tells a full end-to-end story about exactly what takes place and when.

I'll unroll these if it's your preference, but to be frank, I find the existing tests to be pretty difficult to sift through and reason about. I find a test matrix of conditions and expected states much much easier to reason through and write comprehensive test case coverage (not code coverage) with. I also find it much much easier to reason about cache behaviour against a more real-world like "server". The transient nocks make the tests more fragile, and make test failures harder to diagnose. For example: if the caching behaviour breaks and it gets a miss when the tests expect a hit, then that extra hit to the nock will completely break everything. Best case, there's no nocked request left and the error is somewhat clear. Worst case, there's more nocked requests and so the cache hit that should have been a miss gets whatever the next response is. Responses are now desynced from their intended requests. This leads to incredibly misleading test results very quickly. The tests themselves also already check which requests are made using `x-cache-status` and company. As long as there's some request-counting nocks to test the veracity of `x-cache-status`, then it seems trustworthy enough to me. Maybe a combined approach is to use persistent nocks but track request counts made to each and `t.equal()` all of them at the end.

nlf commented:
> > i do think this deserves a closer look to try to determine how we're ending up with `content-encoding: null` in entries in the first place. i think it likely is the code that blindly copies the header when `compress` is `false`, which should check to make sure the header is present before copying.
> 
> Indeed. I addressed that in my previous version of this PR (see [ef415a9](https://github.com/npm/make-fetch-happen/commit/ef415a9e4124c82165dcdd2413a9b0071ddfe385)), but later I concluded the Content-Encoding handling was really a red herring for the underlying issue of not varying on `compress`.

I agree, I was more just questioning if there were any code paths that could cause that `null` header. I don't think there are, but it felt worth raising the question.

> > however, we should _not_ store the `content-encoding` if we've stored a decompressed entry as it is a lie and we know that. it should only be stored if `compress` is `false` and the header exists on the response.
> 
> I agree on principle, but not sure it's that simple in practice, 👇
> 
> > arguably, minipass-fetch is what should be removing that header from the response when it knows it'll be decompressing the body, but we can certainly guard against it here too. sending a decompressed response with a `content-encoding` header signaling otherwise is flat out wrong.
> 
> I agree that minipass-fetch is the ideal place to remove the Content-Encoding header and that barring that make-fetch-happen could almost always detect when it's a lie. However, minipass-fetch's behaviour of decompressing the body but leaving the Content-Encoding header matches that of node-fetch. That behaviour compat may be important for minipass-fetch and/or make-fetch-happen as they aim to be drop-in replacements for node-fetch?
> 
> Notably, the browser `fetch()` (at least in recent Firefox) does the right thing and removes the Content-Encoding header if it was able to decode the body (and leaves it if not).

I think the browser behavior is the most sensible. I don't think maintaining exact feature parity with `node-fetch` is critical, and I'd prefer to keep up with documented standards rather than another implementation of an interpretation of the same spec, if that makes sense 😅 

That said, I'll do some digging and see if there's anything from the IETF clarifying how things should behave. In the mean time, I think we can fix it here and if/when it lands in `minipass-fetch` we can remove the workaround.

> 
> > also, since we know that the failure state here is we end up with a cache entry that has `content-encoding: null` in its headers, we should extend the `validateEntry` function [here](https://github.com/npm/make-fetch-happen/blob/main/lib/cache/entry.js#L145-L151) such that it considers any entry in that state to be invalid, which will remove them from the index and clean up the bad entries on the fly.
> 
> Ah, great point! Will add this.
> 
> I guess the same mechanism could be used to prevent the [incorrect behaviour continuing to happen for existing caches](https://github.com/npm/make-fetch-happen/pull/65#discussion_r762171534), but the effect would be to wipe the cache and maybe that's too dramatic?

Yeah, I think wiping the cache is a bit much. Maybe a fair balance is to make the `revalidate` method set the compress option. The tests for `compress` in the cache policy can be tweaked so if the compress option is missing it uses the presence of a `content-encoding` header on the response as an implied `{ compress: true }`, since that's what the old behavior would be. Do you think that works?

> 
> > and lastly, while i appreciate that you've written tests for these things, i think the DRY approach there leaves the tests in a state that each one doesn't tell a full story. for tests, it's preferable to keep as much context as is reasonable within the test. persistent nocks don't tell us how many times we expect the actual network request to be made. a wrapper function for all of the assertions takes a lot of context away from the test and requires us to scroll down and parse the wrapper function against the inputs. i'd love it if we can unroll those abstractions so that each test tells a full end-to-end story about exactly what takes place and when.
> 
> I'll unroll these if it's your preference, but to be frank, I find the existing tests to be pretty difficult to sift through and reason about. I find a test matrix of conditions and expected states much much easier to reason through and write comprehensive test case coverage (not code coverage) with. I also find it much much easier to reason about cache behaviour against a more real-world like "server". The transient nocks make the tests more fragile, and make test failures harder to diagnose. For example: if the caching behaviour breaks and it gets a miss when the tests expect a hit, then that extra hit to the nock will completely break everything. Best case, there's no nocked request left and the error is somewhat clear. Worst case, there's more nocked requests and so the cache hit that should have been a miss gets whatever the next response is. Responses are now desynced from their intended requests. This leads to incredibly misleading test results very quickly. The tests themselves also already check which requests are made using `x-cache-status` and company. As long as there's some request-counting nocks to test the veracity of `x-cache-status`, then it seems trustworthy enough to me. Maybe a combined approach is to use persistent nocks but track request counts made to each and `t.equal()` all of them at the end.

Absolutely fair sentiment, and it definitely is a matter of personal preference. How's this for a fair balance, instead of calling `makeRequest` and that function doing a loop internally and changing its assertions based on count, we do 3 separate calls to `makeRequest` that pass in the expected values for the assertions. Something like:


```js
t.test('separate caches', async (t) => {
  setupNock(t)

  const cachePath = t.testdir()

  await assertRequest(t, { cachePath, compress: false }, { status: 'miss', body: CONTENT_GZIP, encoding: 'gzip' })
  await assertRequest(t, { cachePath, compress: false }, { status: 'hit', body: CONTENT_GZIP, encoding: 'gzip' })
  await assertRequest(t, { cachePath, compress: false }, { status: 'hit', body: CONTENT_GZIP, encoding: 'gzip' })

  await assertRequest(t, { cachePath, compress: true }, { status: 'miss', body: CONTENT, encoding: null })
  await assertRequest(t, { cachePath, compress: true }, { status: 'hit', body: CONTENT, encoding: null })
  await assertRequest(t, { cachePath, compress: true }, { status: 'hit', body: CONTENT, encoding: null })
})
```

I think that combined with a way to get a request count out of `nock` like you mentioned would satisfy everything I'd like to see, while being a lot less noisy than fully unrolling. Is that a fair compromise?

tsibley commented:
'nother round of responses below, but let me preface by saying thanks for all your thinking and time on this issue so far!  :pray:

> I was more just questioning if there were any code paths that could cause that `null` header. I don't think there are, but it felt worth raising the question.

Ah, hmm. Yes, good point. The other place it happens still is here:

https://github.com/npm/make-fetch-happen/blob/c72173a046f9aa412e233159f89cabb9b6a5b9f6/lib/cache/entry.js#L427-L430

when the existing (buggy) cached metadata contains `content-encoding: null` but the new revalidated response metadata doesn't have the `content-encoding` key, then `null` will be preserved. It won't happen when starting from an empty cache with the code in this PR.

I think this is fixed by the `validateEntry` change you requested above to purge such existing cache entries.  But maybe also good to guard against similar future bugs by checking that the cached value is truthy (like `getMetadata()` does)?

> rather than another implementation of an interpretation of the same spec, if that makes sense :sweat_smile:

:-) sure does!

> That said, I'll do some digging and see if there's anything from the IETF clarifying how things should behave. In the mean time, I think we can fix it here and if/when it lands in minipass-fetch we can remove the workaround.

👍 I'll try to make this change.

> Maybe a fair balance is to make the `revalidate` method set the compress option. The tests for `compress` in the cache policy can be tweaked so if the compress option is missing it uses the presence of a `content-encoding` header on the response as an implied `{ compress: true }`, since that's what the old behavior would be. Do you think that works?

Hmm, I'm not sure I follow.  Here's my understanding:

With the old behaviour, `content-encoding` can be present in the cache entry metadata regardless of the `compress` option, with either a good value or a bad value.  (Explanation in the last paragraph of ad58a8e8ba99187b3b6459c41c0c95c5e935fb79.)  It has the value `null` in two circumstances:

1. `compress: false` in the request options, a cache miss, and no `content-encoding` header on the response, [set in `getMetadata()`](https://github.com/npm/make-fetch-happen/blob/ff3d03c7828bb029de556c5df1714e30c173c81e/lib/cache/entry.js#L101-L104).

2. `compress: true` in the request options, a cache revalidation of the stored result of (1), and no `content-encoding` header on the response to the revalidation request, [set in `revalidate()`](https://github.com/npm/make-fetch-happen/blob/ff3d03c7828bb029de556c5df1714e30c173c81e/lib/cache/entry.js#L427-L430)

So I'm not sure how to "rescue" existing cache entries that don't distinguish the value of `compress`. I'm not sure it's actually possible. Since a cache is fundamentally a transient thing anyway, it seems acceptable to me to write in the release notes something like, "If you care about this bug, then blow away your cache to prevent it from reoccurring; otherwise, feel free to keep using your cache."

> I think that combined with a way to get a request count out of `nock` like you mentioned would satisfy everything I'd like to see, while being a lot less noisy than fully unrolling. Is that a fair compromise?

Yep, I'll give that a try and see how it goes!

tsibley commented:
Todos for me from review above:

- [x] Remove unnecessary `|| {}`
- [x] Compare to `_req.compress` instead of `request.compress` for consistency 
- [ ] Update `validateEntry()` to return false when `content-encoding: null` is in the metadata
- [ ] Maybe condition `revalidate()`'s response metadata header copying on truthy values?
- [ ] Partially unroll/refactor new tests
- [ ] Assert request counts to persistent nocks in new tests

nlf commented:
> With the old behaviour, content-encoding can be present in the cache entry metadata regardless of the compress option, with either a good value or a bad value. (Explanation in the last paragraph of ad58a8e.) It has the value null in two circumstances:

Ahhhh ok, this was my misunderstanding. I was thinking we were only storing the `content-encoding` if `compress` was `false`, but if that's not the case that won't work. What about `accept-encoding`? is that only stored when `compress` is `false`?

tsibley commented:
> Ahhhh ok, this was my misunderstanding. I was thinking we were only storing the `content-encoding` if `compress` was `false`, but if that's not the case that won't work. What about `accept-encoding`? is that only stored when `compress` is `false`?

Unfortunately not. When present, it was always stored regardless of `compress`. When absent, it was stored as `null` when `compress: false`.

nlf commented:
> Unfortunately not. When present, it was always stored regardless of `compress`. When absent, it was stored as `null` when `compress: false`.

Welp. So that leaves us with a couple of options:

1. We consider anything lacking the `compress` option in its metadata invalid. This has the consequence of effectively clearing the entire cache.
2. When we encounter an entry that lacks a `compress` option, we read the first few bytes of the content the entry points to and try to determine if it's compressed data or not. This is a non-trivial amount of work, and I'm not sure that the trade off of reading some content for every existing cache entry is much better than just sending the request again. It's also very significantly more complicated.

I think we go with option 1, unless you have any better suggestions?

Also, I just want to reiterate: _thank you_, again, for all of your time and effort you've put into this

nlf commented on lib/cache/entry.js:
```diff
@@ -143,6 +138,12 @@ class CacheEntry {
+              entry.metadata.resHeaders &&
+              entry.metadata.resHeaders['content-encoding'] === null)
+            return false
+
```
kind of a nitpick, but what do you think about making this use `||` instead of `&&` between each check? that way if, somehow, someone ends up with an entry that doesn't have `metadata` or `resHeaders` that would also get cleared out as invalid

tsibley replied:
Hmm, this is what you mean?

```js
          if (!entry.metadata ||
              !entry.metadata.resHeaders ||
              entry.metadata.resHeaders['content-encoding'] === null)
```

That sounds reasonable, but I think we'll have to update a few tests to more thoroughly mock cache entries. If I remember correctly, I had initially just had the `=== null` check, but that ran afoul of tests which omitted `metadata.resHeaders`, so I added the other checks.

tsibley replied:
I guess if we start going down that route, it could make sense to more systematically validate everything the code assumes to be in the entry, like `reqHeaders` and `url` and stuff too? If that's the case, then I think that'd be better in a separate PR.

nlf replied:
you're absolutely right, i think more actual validation is what we really want, and it does make way more sense in a separate PR

tsibley commented:
Agreed that option 2 is probably not the best set of tradeoffs. Option 1 sounds ok to me, and I can add that validation check if we chose that route.

From my perspective, I think there's also an Option 0 which is do nothing technically to clear the cache, but warn users in release notes that until they clear their cache, they may still experience the bug around `compress`, particularly if they're making requests to the same URLs with different `compress` values. The only upside here is that anyone who doesn't have the bug because of their use case can keep their cache.

nlf commented:
> From my perspective, I think there's also an Option 0 which is do nothing technically to clear the cache, but warn users in release notes that until they clear their cache, they may still experience the bug around `compress`, particularly if they're making requests to the same URLs with different `compress` values. The only upside here is that anyone who doesn't have the bug because of their use case can keep their cache.

This makes sense. I think the majority of users are by way of the `npm` cli, which only ever makes requests with `{ compress: true }`, so for the majority of people this won't be an issue. When we revalidate requests, we do end up storing the compress option right? If that's the case, it kinda becomes self-healing.

tsibley commented:
> so for the majority of people this won't be an issue.

Excellent!

> When we revalidate requests, we do end up storing the compress option right? If that's the case, it kinda becomes self-healing.

Yes and no, unfortunately. The current `compress` value during revalidation is indeed stored back, but it doesn't necessarily heal the cache entry. If the value of `compress` during revalidation matches the value of `compress` used to make the original request for the cache entry, then the entry is indeed "healed" from future cache mishits. But if the two `compress` values don't match, then the cache mishit is persisted into the entry and can occur again.

nlf commented:
> But if the two compress values don't match, then the cache mishit is persisted into the entry and can occur again.

I see, so this means if I have an entry that doesn't have the compress option at all, and I make a request with `{ compress: false }` we'll respond with whatever we have cached and then update the cache index to hold `{ compress: false }` even in the case where the content was originally fetched with `{ compress: true }` since we have no idea that's how it was fetched.

I think this is reasonable. Unless you have strong feelings that it should be fixed, I'm comfortable calling this good to go!

tsibley commented:
> … since we have no idea that's how it was fetched.

Precisely.

Agreed that it feels reasonable. If there's nothing else outstanding, then I'm happy to see this merged whenever you're ready. :-)

wraithgar commented:
Thanks for your patience here @tsibley

tsibley commented:
No problem. Thanks for merging! 🎉
